File: src/ui/panels/display.ts
import type { DevicePanel, DeviceSession } from "../../devices/panel.ts";
import { el } from "../dom.ts";
import { panel, statusPill } from "../panel.ts";

export interface DisplayControl {
  kind: "select" | "range" | "toggle";
  command: string;
  label: string;
  title?: string;
  value: string | number | boolean;
  min?: number;
  max?: number;
  options?: { value: string; label: string }[];
}

export interface DisplayState {
  label: string;
  width: number;
  height: number;
  /** Base64 of the framebuffer, in the panel's own page-major layout. */
  frame: string;
  pattern: string;
  patterns: { key: string; label: string; title: string }[];
  lastShowMs: number;
  bytesSent: number;
  controls: DisplayControl[];
  details: { label: string; value: string; title?: string }[];
  previewIsSent: boolean;
}

/** Pixels per display pixel in the preview. */
const ZOOM = 3;

/**
 * One panel for every small framebuffer display.
 *
 * There is no polling here, and that is not an oversight. These parts are
 * write-only and hold no state worth reading back, so a timer would spend the
 * bus on nothing -- a full 128x64 frame is over a kilobyte and eighteen chunks.
 * The panel draws when asked and otherwise leaves the bus alone.
 *
 * The preview is labelled as what was *sent* rather than what is shown, every
 * time, because that distinction is the whole difficulty of working with these
 * modules: a dark screen beside a perfect preview is the normal failure, and a
 * picture on a page is a very easy thing to mistake for a screenshot.
 */
export class DisplayPanel implements DevicePanel {
  readonly root: HTMLElement;
  readonly #session: DeviceSession;
  readonly #status = statusPill("Ready", "ok");
  readonly #canvas = el("canvas", { class: "screen" }) as HTMLCanvasElement;
  readonly #patterns = el("div", { class: "screen-patterns" });
  readonly #controls = el("div", { class: "lux-controls" });
  readonly #facts = el("dl", { class: "facts" });
  #built = false;

  constructor(session: DeviceSession) {
    this.#session = session;
    const p = panel("Screen");
    this.root = p.root;
    p.actions.append(this.#status.node);

    p.body.append(
      el("div", { class: "screen-columns" }, [
        el("div", { class: "screen-side" }, [
          el("p", { class: "lux-fill-label", text: "Sent to the panel" }),
          el("div", { class: "screen-frame" }, [this.#canvas]),
          el("p", {
            class: "caption",
            text:
              "This is the host's copy of what was sent, not a picture of the " +
              "screen. The interface is write-only — nothing can be read back — " +
              "so a dark panel beside a correct preview here is the normal way " +
              "these modules fail, and usually means the charge pump.",
          }),
        ]),
        el("div", {}, [
          el("p", { class: "lux-fill-label", text: "Test patterns" }),
          this.#patterns,
          this.#controls,
          this.#facts,
        ]),
      ]),
    );
  }

  show(): void {
    void this.#refresh();
  }

  hide(): void {}

  async #refresh(): Promise<void> {
    try {
      this.#render(await this.#session.poll<DisplayState>());
    } catch (err) {
      this.#status.set(err instanceof Error ? err.message : String(err), "error");
    }
  }

  async #run(command: string, args: unknown[], busy: string): Promise<void> {
    this.#status.set(busy, "busy");
    try {
      this.#render(await this.#session.command<DisplayState>(command, ...args));
    } catch (err) {
      this.#status.set(err instanceof Error ? err.message : String(err), "error");
    }
  }

  #render(state: DisplayState): void {
    this.#build(state);
    this.#status.set(
      state.lastShowMs
        ? `${state.bytesSent} B in ${state.lastShowMs.toFixed(0)} ms`
        : "ready",
      "ok",
    );
    this.#paint(state);

    const rows = [
      {
        label: "Panel",
        value: `${state.width} × ${state.height}`,
      },
      {
        label: "Last transfer",
        value: state.lastShowMs
          ? `${state.bytesSent} bytes · ${state.lastShowMs.toFixed(0)} ms`
          : "—",
        title:
          "A whole frame every time. These controllers take no partial update " +
          "over I²C without setting a narrower window first, so a redraw is " +
          "the full buffer plus the commands that address it.",
      },
      ...state.details,
    ];
    this.#facts.replaceChildren(
      ...rows.flatMap((row) => [
        el("dt", { text: row.label, title: row.title ?? "" }),
        el("dd", { text: row.value, title: row.title ?? "" }),
      ]),
    );
  }

  /**
   * Decode the framebuffer and draw it. The layout is the panel's own: one
   * byte is eight *vertical* pixels, so the bit index is the row within a page.
   */
  #paint(state: DisplayState): void {
    const bytes = Uint8Array.from(atob(state.frame), (c) => c.charCodeAt(0));
    this.#canvas.width = state.width * ZOOM;
    this.#canvas.height = state.height * ZOOM;

    const context = this.#canvas.getContext("2d");
    if (!context) return;
    const style = getComputedStyle(this.root);
    context.fillStyle = style.getPropertyValue("--screen-off") || "#101314";
    context.fillRect(0, 0, this.#canvas.width, this.#canvas.height);
    context.fillStyle = style.getPropertyValue("--screen-on") || "#8fd3e8";

    for (let page = 0; page < state.height / 8; page++) {
      for (let x = 0; x < state.width; x++) {
        const byte = bytes[page * state.width + x] ?? 0;
        for (let bit = 0; bit < 8; bit++) {
          if ((byte >> bit) & 1) {
            context.fillRect(x * ZOOM, (page * 8 + bit) * ZOOM, ZOOM - 1, ZOOM - 1);
          }
        }
      }
    }
  }

  #build(state: DisplayState): void {
    if (this.#built) {
      this.#syncControls(state);
      return;
    }
    this.#built = true;

    this.#patterns.replaceChildren(
      ...state.patterns.map((pattern) => {
        const button = el("button", { text: pattern.label, title: pattern.title });
        button.addEventListener("click", () => {
          button.disabled = true;
          void this.#run("pattern", [pattern.key], `drawing ${pattern.label}…`).finally(
            () => {
              button.disabled = false;
            },
          );
        });
        return button;
      }),
    );

    this.#controls.replaceChildren(...state.controls.map((c) => this.#control(c)));
  }

  #control(control: DisplayControl): HTMLElement {
    if (control.kind === "toggle") {
      const input = el("input", { type: "checkbox" }) as HTMLInputElement;
      input.checked = Boolean(control.value);
      input.dataset.command = control.command;
      input.addEventListener("change", () => {
        void this.#run(control.command, [input.checked], "…");
      });
      return el("label", { title: control.title ?? "" }, [
        input,
        el("span", { text: control.label }),
      ]);
    }

    if (control.kind === "range") {
      const input = el("input", {
        type: "range",
        min: String(control.min ?? 0),
        max: String(control.max ?? 255),
      }) as HTMLInputElement;
      input.value = String(control.value);
      input.dataset.command = control.command;
      // On change, not input: each move is an I²C write, and a slider dragged
      // across its range would otherwise queue a hundred of them.
      input.addEventListener("change", () => {
        void this.#run(control.command, [Number(input.value)], "…");
      });
      return el("label", { title: control.title ?? "" }, [
        el("span", { text: control.label }),
        input,
      ]);
    }

    const select = el("select", { title: control.title ?? "" }) as HTMLSelectElement;
    for (const option of control.options ?? []) {
      select.append(el("option", { value: option.value, text: option.label }));
    }
    select.value = String(control.value);
    select.dataset.command = control.command;
    select.addEventListener("change", () => {
      void this.#run(control.command, [select.value], "reconfiguring…");
    });
    return el("label", { title: control.title ?? "" }, [
      el("span", { text: control.label }),
      select,
    ]);
  }

  #syncControls(state: DisplayState): void {
    for (const control of state.controls) {
      const node = this.#controls.querySelector<HTMLInputElement | HTMLSelectElement>(
        `[data-command="${control.command}"]`,
      );
      if (!node || document.activeElement === node) continue;
      if (node instanceof HTMLInputElement && node.type === "checkbox") {
        node.checked = Boolean(control.value);
      } else {
        node.value = String(control.value);
      }
    }
  }
}

File: python/webblinka/drivers/display.py
"""Shared ground for small framebuffer displays.

Every one of these parts works the same way: a block of memory on the host is
drawn into with the same handful of primitives, then shipped to the panel whole.
So the patterns live here once, written against the ``framebuf`` API that every
Adafruit mono display shares, and a driver supplies only the part that differs
-- how to make the device, how big it is, and which of its registers are worth
exposing.

**A display is write-only, and that shapes the whole panel.** Over I2C the
SSD1306 has no read path: you cannot ask it what is on the screen, what size it
thinks it is, or even whether it is there beyond the address ACK. Every command
is fire-and-forget. So the preview this driver returns is emphatically not a
screenshot -- it is the host's copy of what was *sent*, and a screen that stays
dark while the preview looks perfect is the normal way this hardware fails.

That is exactly why the patterns are worth having, and why they are these
patterns. Each one is chosen so that its *appearance* names the fault:

* nothing at all, on a bus that ACKed every byte, is the charge pump -- these
  panels generate their own 7.5 V and are simply dark without it;
* a border missing its bottom edge is the wrong height, 32 driven as 64;
* page stripes that come out uneven are a page-addressing fault;
* and a ruler whose ends wrap is a column offset -- the SH1106 that gets sold
  as an SSD1306 has 132 columns of RAM and starts two in, so an SSD1306 driver
  puts everything two pixels across and wraps the edge.

None of that is visible in a register. It is only visible on the glass, which
is why the answer to "is it working" is a pattern and not a status byte.
"""

from __future__ import annotations

import base64
import time
from typing import Any

from ..font5x8 import ensure_font
from .base import Driver

#: What the panel offers, in the order it offers it. The first two answer "is
#: it alive at all"; the rest answer "is it the geometry I told it".
PATTERNS: list[dict[str, str]] = [
    {
        "key": "all_on",
        "label": "All on",
        "title": (
            "Every pixel lit. The first thing to try: a panel that stays dark "
            "through this, on a bus that acknowledged every byte, is almost "
            "always the charge pump — these generate their own 7.5 V and show "
            "nothing without it."
        ),
    },
    {
        "key": "all_off",
        "label": "All off",
        "title": "Every pixel dark. Finds pixels stuck on, and clears the glass.",
    },
    {
        "key": "border",
        "label": "Border",
        "title": (
            "A one-pixel frame around the whole panel. It is the geometry test: "
            "if the bottom edge is missing you are driving a 32-row panel as 64, "
            "and if an edge is cut off the width is wrong."
        ),
    },
    {
        "key": "pages",
        "label": "Page stripes",
        "title": (
            "Alternating eight-pixel bands, which is how the memory is actually "
            "organised — one byte is eight vertical pixels. Uneven bands mean a "
            "page addressing fault rather than a drawing one."
        ),
    },
    {
        "key": "checker",
        "label": "Checkerboard",
        "title": (
            "Single-pixel checks: the hardest thing for the charge pump to drive "
            "and the easiest place to see ghosting between neighbouring pixels."
        ),
    },
    {
        "key": "ruler",
        "label": "Column ruler",
        "title": (
            "Ticks every 8 columns, taller every 32, with both end columns "
            "marked. If the ends wrap around, the controller is an SH1106 rather "
            "than an SSD1306 — it has 132 columns of RAM and starts two in, so "
            "everything lands two pixels across."
        ),
    },
    {
        "key": "text",
        "label": "Text",
        "title": "The built-in 5×8 font, with the panel's own geometry written on it.",
    },
]


class MonoDisplay(Driver):
    """Base for one-bit-per-pixel framebuffer displays."""

    LABEL = "Display"

    def __init__(self, bus, address: int) -> None:
        super().__init__(bus, address)
        self._device = None
        self._pattern = ""
        self._last_show_ms = 0.0
        self._bytes_sent = 0

    # -- what a subclass supplies -----------------------------------------

    def make_device(self):
        """Construct and initialise the display. Returns the framebuf object."""
        raise NotImplementedError

    def geometry(self) -> dict[str, int]:
        """``width`` and ``height`` in pixels."""
        raise NotImplementedError

    def frame_bytes(self) -> bytes:
        """The host's copy of the pixels, in the panel's own memory layout."""
        raise NotImplementedError

    def controls(self) -> list[dict[str, Any]]:
        return []

    def details(self) -> list[dict[str, Any]]:
        return []

    # -- the shared work ---------------------------------------------------

    def start(self) -> dict[str, Any]:
        self._device = self.make_device()
        return {"address": self.address, "label": self.LABEL, **self.geometry()}

    def stop(self) -> None:
        self._device = None

    def command(self, name: str, args: list[Any]) -> Any:
        if name == "pattern":
            return self.draw(str(args[0]))
        return super().command(name, args)

    def draw(self, key: str) -> dict[str, Any]:
        device = self.require()
        width = self.geometry()["width"]
        height = self.geometry()["height"]

        device.fill(0)
        if key == "all_on":
            device.fill(1)
        elif key == "all_off":
            pass
        elif key == "border":
            device.rect(0, 0, width, height, 1)
        elif key == "pages":
            for page in range(0, height, 8):
                if (page // 8) % 2 == 0:
                    device.fill_rect(0, page, width, 8, 1)
        elif key == "checker":
            for y in range(height):
                for x in range(y % 2, width, 2):
                    device.pixel(x, y, 1)
        elif key == "ruler":
            _ruler(device, width, height)
        elif key == "text":
            # The font has to be passed explicitly. framebuf's default resolves
            # font5x8.bin relative to the working directory, which is not
            # something a driver should be depending on.
            font = ensure_font()
            device.text(f"{width}x{height}", 0, 0, 1, font_name=font)
            device.text(f"{self.LABEL}", 0, 10, 1, font_name=font)
            device.text("0123456789ABCDEF", 0, 20, 1, font_name=font)
        else:
            raise LookupError(f"no pattern {key!r}")

        self._pattern = key
        self.show()
        return self.poll()

    def show(self) -> None:
        """Ship the framebuffer and record what it cost.

        Worth recording because it is the whole expense of using one of these.
        A 128x64 panel is 1025 bytes with the data prefix, which Blinka chops
        into 60-byte chunks -- eighteen of them, plus the six commands that set
        the addressing window before it. That is not a poll, it is a transfer,
        and it is why this panel does not refresh on a timer like the others.
        """
        device = self.require()
        began = time.monotonic()
        device.show()
        self._last_show_ms = (time.monotonic() - began) * 1000
        self._bytes_sent = len(self.frame_bytes())

    def poll(self) -> dict[str, Any]:
        frame = self.frame_bytes()
        return {
            "label": self.LABEL,
            **self.geometry(),
            # Base64 rather than a list of a thousand integers: the same bytes
            # in a third of the JSON, and the panel has atob built in.
            "frame": base64.b64encode(frame).decode("ascii"),
            "pattern": self._pattern,
            "patterns": PATTERNS,
            "lastShowMs": self._last_show_ms,
            "bytesSent": self._bytes_sent,
            "controls": self.controls(),
            "details": self.details(),
            # Said on every reading, because it is the one thing about this
            # panel that is easy to forget while looking at a picture of it.
            "previewIsSent": True,
        }

    def require(self):
        if self._device is None:
            raise RuntimeError(f"{self.LABEL} not started")
        return self._device


def _ruler(device, width: int, height: int) -> None:
    """Ticks along the top, and both end columns marked full height.

    The end columns are the point. A controller whose RAM is wider than its
    glass -- the SH1106 is 132 columns against 128 of screen -- puts them
    somewhere other than the edges, and nothing else in a normal image makes
    that visible.
    """
    device.vline(0, 0, height, 1)
    device.vline(width - 1, 0, height, 1)
    for x in range(0, width, 8):
        tick = 8 if x % 32 == 0 else 4
        device.vline(x, 0, tick, 1)
        device.vline(x, height - tick, tick, 1)
